[HLSL][SPIRV]Add SPIRV generation for HLSL dot

pow2clk opened the pull request:
This adds the SPIRV fdot, sdot, and udot intrinsics and allows them to be created at codegen depending on the target architecture. This required moving some of the DXIL-specific choices to DXIL instruction expansion out of codegen and providing it with at a more generic fdot intrinsic as well.

Removed some stale comments that gave the obsolete impression that type conversions should be expected to match overloads.

The SPIRV intrinsic handling involves generating multiply and add operations for integers and the existing OpDot operation for floating point.

New tests for generating SPIRV float and integer dot intrinsics are added as well.

Incidentally changed existing dot intrinsic definitions to use DefaultAttrsIntrinsic to match the newly added inrinsics

Fixes #88056

Commit:
Add SPIRV generation for HLSL dot

This adds the SPIRV fdot, sdot, and udot intrinsics and allows
them to be created at codegen depending on the target architecture.
This required moving some of the DXIL-specific choices to DXIL
instruction expansion out of codegen and providing it with at a
more generic fdot intrinsic as well.

Removed some stale comments that gave the obsolete impression that
type conversions should be expected to match overloads.

The SPIRV intrinsic handling involves generating multiply and add
operations for integers and the existing OpDot operation for
floating point.

New tests for generating SPIRV float and integer dot intrinsics are
added as well.

Incidentally changed existing dot intrinsic definitions to use
DefaultAttrsIntrinsic to match the newly added inrinsics

Fixes #88056

Commit:
Correct fdot test RUN line

farzonl commented on clang/test/CodeGenHLSL/builtins/dot.hlsl:
```diff
@@ -7,155 +7,155 @@
+// NATIVE_HALF: ret i16 %hlsl.dot
 int16_t test_dot_short2(int16_t2 p0, int16_t2 p1) { return dot(p0, p1); }
 
-// NATIVE_HALF: %dx.dot = call i16 @llvm.dx.sdot.v3i16(<3 x i16> %0, <3 x i16> %1)
```
i'd like to see some spv tests so we can see `@llvm.spv.sdot`, `@llvm.spv.udot`, and `@llvm.spv.fdot` get emitted\tested.

farzonl replied:
if its cleaner to do this in another file im fine with that too.

pow2clk replied:
Done

farzonl commented on llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:
```diff
@@ -68,28 +69,65 @@ static Value *expandAbs(CallInst *Orig) {
-    Elt0 = Builder.CreateExtractElement(A, I);
-    Elt1 = Builder.CreateExtractElement(B, I);
+  Result = Builder.CreateMul(Elt0, Elt1);
+  for (unsigned i = 1; i < AVec->getNumElements(); i++) {
```
this needs to be capitalizes to abide by the llvm style guide.

pow2clk replied:
I try so hard to be agnostic about style guides, but stuff like this challenge my lack of creed. 😣

pow2clk replied:
Done

farzonl commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1446,6 +1449,67 @@ bool SPIRVInstructionSelector::selectRsqrt(Register ResVReg,
       .constrainAllUses(TII, TRI, RBI);
 }
 
+// Since there is no integer dot implementation, expand by piecewise multiplying
```
```suggestion
// Since there is no integer dot implementation until spirv 1.6, expand by piecewise multiplying
```

pow2clk replied:
Done

farzonl commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -2157,6 +2221,16 @@ bool SPIRVInstructionSelector::selectIntrinsic(Register ResVReg,
   case Intrinsic::spv_thread_id:
     return selectSpvThreadId(ResVReg, ResType, I);
+  case Intrinsic::spv_fdot:
+    return BuildMI(BB, I, I.getDebugLoc(), TII.get(SPIRV::OpDot))
```
Maybe we should do a `selectDot` function? Would be more consistent with how all the other cases are setup. Also you are doing operand checks before addUse in `selectIntegerDot` that you aren't doing here.

pow2clk replied:
I don't mind creating a float function to expand the fdot. I don't think you're suggesting that I have one call for float and integer as that's the opposite of what you suggested [here](https://github.com/llvm/llvm-project/pull/102872#discussion_r1714053193). Are the checks you're referring to the asserts?

farzonl replied:
no i'm saying create a function so that its cleaner to asserts on the number of operands and the validity of those operands.

pow2clk replied:
Done assuming I've understood the suggestion correctly.

farzonl commented:
could you fixup normalize to use some of you fdot expansion? I think you need to split the intrinsic check from the emitting check 
 https://github.com/llvm/llvm-project/blob/9cf27a4d8b1da0e7b51eacb9fb6096155c294d3f/llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp#L278-L289

Commit:
Repond to feedback

Add SPIRV checks to dot.hlsl codegen test. This required some
reorganizing of the test to capture the interchange format early
and removing the dependence on argument position

Reluctantly conform iterator variable to style guide

Add float dot selector for SPIRV with asserts

Commit:
Leverage dot product expansion in normalize expansion

Since normalize requires a dot product operation, changing the signature
so that it can be called from within another operation expansion allows
sharing of code.

pow2clk commented:
> could you fixup normalize to use the fdot expansion?
> 
> https://github.com/llvm/llvm-project/blob/9cf27a4d8b1da0e7b51eacb9fb6096155c294d3f/llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp#L278-L289

It makes the signature a bit unusual, but it works!

pow2clk commented on clang/test/CodeGenHLSL/builtins/dot.hlsl:
```diff
@@ -1,161 +1,172 @@
-// NATIVE_HALF: ret half %dx.dot
-// NO_HALF: %dx.dot = fmul float %0, %1
-// NO_HALF: ret float %dx.dot
+#ifdef __HLSL_ENABLE_16_BIT
```
I needed somewhere to capture the dx/spv indicator that was run early and every time, so I moved these out of the way.

pow2clk commented on clang/test/CodeGenHLSL/builtins/dot.hlsl:
```diff
@@ -1,161 +1,172 @@
 
-// CHECK: %dx.dot = call i32 @llvm.dx.sdot.v3i32(<3 x i32> %0, <3 x i32> %1)
-// CHECK: ret i32 %dx.dot
+// CHECK: %hlsl.dot = call i32 @llvm.[[ICF]].sdot.v3i32(<3 x i32>
```
I opted to chop off the rest instead of delve into regexpamancy because that's what other tests in this directory have done.

farzonl replied:
hmm i like this `[[ICF]]`, hmm we should probably put something in the backlog to update all the tests to do this.

pow2clk replied:
I filed an issue: #105710

Commit:
Small comment clarifications

farzonl commented on llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:
```diff
@@ -363,9 +382,13 @@ static bool expandIntrinsic(Function &F, CallInst *Orig) {
     break;
+  case Intrinsic::dx_fdot:
+    Result =
+        expandFloatDotIntrinsic(Orig, Orig->getOperand(0), Orig->getOperand(1));
```
you are right that is a little weird maybe, a function overload would be better? Its a nitpick so feel free to ignore.

pow2clk replied:
Done

farzonl commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1446,6 +1452,90 @@ bool SPIRVInstructionSelector::selectRsqrt(Register ResVReg,
+
+  assert(GR.getScalarOrVectorComponentCount(VecType) > 1 &&
+         "dot product requires a vector of at least 2 components");
+
```
should we also add an assert to make sure its a float vector? I know we have guarantees about that via `Intrinsic::spv_fdot` but I was concerned about misuse of the `selectFloatDot` function if say someone else wants to come along and use it for opencl? Its a nit pick so feel free to ignore.

pow2clk replied:
Done

Commit:
Add type check to dot operands and homogenize DXIL expansion call

Commit:
clang-format

Commit:
Merge branch 'main' into spirv_dot
